### ISSUE TITLE
Try cheaper cross-attention mechanisms

### ISSUE THREAD

**Issue by Malikeh97** [Malikeh97/lingua#2]
Trying cheaper cross-attention mechanisms (like linear attention) from (https://github.com/fla-org/flash-linear-attention?tab=readme-ov-file)

**Comment by Malikeh97** [comment]
- Good resource to start with: https://arxiv.org/abs/2510.26692
- A Survey of Linear Attention: Algorithm, Theory, Application, and Infrastructure: https://www.techrxiv.org/users/1020124/articles/1382431-a-survey-of-linear-attention-algorithm-theory-application-and-infrastructure
- Interesting video: https://www.youtube.com/watch?v=pUCWwGR5WmQ

- Olmo Hybrid: From Theory to Practice: https://allenai.org/papers/olmo-hybrid

**Comment by Malikeh97** [comment]
Here is the report for preliminary experiments with Kimi KDA:
https://github.com/Malikeh97/lingua/blob/enc-dec-training/apps/minimal_squad/reports/Mar10th_LINEAR_ATTN_REPORT.md

Summary: 

- Encoder size matters.
- In CEPE-style training with FLA-backed KDA, we observed a 0.6 EM improvement with the 400M ModernBERT encoder and an ~11% improvement with the 150M ModernBERT encoder. In this setting, training was also 1.2–1.3× faster compared to the softmax-based full attention mechanism.
- In contrast, the frozen and full fine-tuning setups were significantly slower than the softmax-based full attention baseline, which requires further investigation.

Overall, results were promising. We can think of linear attention as an option for performance and efficiency improvement.
